File: Algorithms/search_sort/Python/quicksort.py

def parition(arr, left, right):
    ind, key = left, arr[left]

    while left < right:
        while left < len(arr) and arr[left] <= key:
            left += 1
        while right > 0 and arr[right] > key:
            right -= 1
        if left < right:
            arr[left], arr[right] = arr[right], arr[left]

    arr[right], arr[ind] = arr[ind], arr[right]
    return right


def quick(arr, left, right):
    if left < right:
        p = parition(arr, left, right)
        quick(arr, p + 1, right)
        quick(arr, left, p - 1)

File: Algorithms/search_sort/C++/quicksort.cpp
#include <vector>

using namespace std;

void swap(vector<int>&, int, int);
int partition(vector<int>&, int, int);
void sort(vector<int>&, int, int);

void swap(vector<int> &arr, int left, int right){

    int t = arr[left];
    arr[left]  = arr[right];
    arr[right] = t;
}

int partition(vector<int> &arr, int left, int right){

    int pivot = arr[left];
    int ind = left;
    while(left < right){

        while(left < arr.size() && arr[left] <= pivot)
            ++left;
        while(arr[right] > pivot)
            --right;
        if(left < right){
            swap(arr, left, right);
        }
    }
    swap(arr, right, ind);
    return right;

}


void sort(vector<int> &arr, int left, int right){
    if(left < right){

        int p = partition(arr, left, right);
        sort(arr, p + 1, right);
        sort(arr, left, p - 1);
    }
}
